sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveWithCTE.scala

```diff
@@ -37,23 +39,159 @@ object ResolveWithCTE extends Rule[LogicalPlan] {
+              case alias as@ SubqueryAlias(_,
+              columnAlias @ UnresolvedSubqueryColumnAliases(_,
+              Union(Seq(anchor, recursion), false, false))) =>
+                cteDef.copy(child =
+                  alias.copy(child =
+                    columnAlias.copy(child =
+                      UnionLoop(cteDef.id, anchor, transformRefs(recursion)))))
+              // If the recursion is described with an UNION (deduplicating) clause then the
+              // recursive term should not return those rows that have been calculated previously,
+              // and we exclude those rows from the current iteration result.
+              case alias @ SubqueryAlias(_,
+                  Distinct(Union(Seq(anchor, recursion), false, false))) =>
+                cteDef.copy(child =
+                  alias.copy(child =
+                    UnionLoop(cteDef.id,
+                      Distinct(anchor),
+                      Except(
+                        transformRefs(recursion),
+                        UnionLoopRef(cteDef.id, cteDef.output, true),
+                        false))))
+              case alias @ SubqueryAlias(_,
+              columnAlias @ UnresolvedSubqueryColumnAliases(_,
+                  Distinct(Union(Seq(anchor, recursion),
+              false, false)))) =>
+                cteDef.copy(child =
+                  alias.copy(child =
+                    columnAlias.copy(child =
+                      UnionLoop(cteDef.id,
+                        Distinct(anchor),
+                        Except(
+                          transformRefs(recursion),
+                          UnionLoopRef(cteDef.id, cteDef.output, true),
+                          false)))))
+              case _ =>
+                // We do not support cases of sole Union (needs a SubqueryAlias above it), nor
+                // Project (as UnresolvedSubqueryColumnAliases have not been substituted with the
+                // Project yet), leaving us with cases of SubqueryAlias->Union and SubqueryAlias->
+                // UnresolvedSubqueryColumnAliases->Union. The same applies to Distinct Union.
+                cteDef.failAnalysis(
+                    errorClass = "INVALID_RECURSIVE_CTE",
+                    messageParameters = Map.empty)
+                throw QueryCompilationErrors.recursiveCteError()
+            }
+          } else {
+            cteDef
           }
+
+          if (newCTEDef.recursive) {
+            // cteDefMap holds "partially" resolved (only via anchor) CTE definitions in the
+            // recursive case.
+            if (newCTEDef.resolved) {
+              newCTEDef.failAnalysis(
+                errorClass = "INVALID_RECURSIVE_CTE",
+                messageParameters = Map.empty)
+              throw QueryCompilationErrors.recursiveCteError()
+            }
+            if (recursiveAnchorResolved(newCTEDef).isDefined) {
+              cteDefMap.put(newCTEDef.id, newCTEDef)
```

Comment:
shall we update the `cteDefMap` when we match the recursion anchor and replace with `UnionLoop`?